contracts/test/CompoundGovernorCountingFractional.t.sol

```diff
@@ -0,0 +1,356 @@
+        Proposal memory _proposal = _buildAnEmptyProposal();
+        uint256 _proposalId = _getProposalId(_proposal);
+        _submitProposal(_proposer, _proposal);
+
+        address _delegate = _majorDelegates[_voterIndex];
+        uint256 _votes = governor.getVotes(_delegate, vm.getBlockNumber() - 1);
+        _forVotes = bound(_forVotes, 0, _votes);
+        _againstVotes = bound(_againstVotes, 0, _votes - _forVotes);
+        _abstainVotes = bound(_abstainVotes, 0, _votes - _forVotes - _againstVotes);
+
+        vm.prank(_delegate);
+        bytes memory _params = abi.encodePacked(uint128(_againstVotes), uint128(_forVotes), uint128(_abstainVotes));
+        governor.castVoteWithReasonAndParams(_proposalId, VOTE_TYPE_FRACTIONAL, "MyReason", _params);
+
+        assertEq(governor.usedVotes(_proposalId, _delegate), _forVotes + _againstVotes + _abstainVotes);
+    }
+}
+
+abstract contract CastVoteWithReasonAndParams is ProposalTest {
+    function testFuzz_CastVotesViaFlexibleVoting(
+        uint256 _proposerIndex,
+        uint256 _voterIndex,
+        uint256 _forVotes,
+        uint256 _againstVotes,
+        uint256 _abstainVotes
+    ) public {
+        _proposerIndex = bound(_proposerIndex, 0, _majorDelegates.length - 1);
+        _voterIndex = bound(_voterIndex, 0, _majorDelegates.length - 1);
+
+        address _proposer = _majorDelegates[_proposerIndex];
+        Proposal memory _proposal = _buildAnEmptyProposal();
+        uint256 _proposalId = _getProposalId(_proposal);
+        _submitProposal(_proposer, _proposal);
+
+        address _delegate = _majorDelegates[_voterIndex];
+        uint256 _votes = governor.getVotes(_delegate, vm.getBlockNumber() - 1);
+        _forVotes = bound(_forVotes, 0, _votes);
+        _againstVotes = bound(_againstVotes, 0, _votes - _forVotes);
+        _abstainVotes = bound(_abstainVotes, 0, _votes - _forVotes - _againstVotes);
+
+        vm.prank(_delegate);
+        bytes memory _params = abi.encodePacked(uint128(_againstVotes), uint128(_forVotes), uint128(_abstainVotes));
+        governor.castVoteWithReasonAndParams(_proposalId, VOTE_TYPE_FRACTIONAL, "MyReason", _params);
+
+        (uint256 _againstVotesCast, uint256 _forVotesCast, uint256 _abstainVotesCast) =
+            governor.proposalVotes(_proposalId);
+
+        assertEq(_againstVotesCast, _againstVotes);
+        assertEq(_forVotesCast, _forVotes);
+        assertEq(_abstainVotesCast, _abstainVotes);
+    }
+
+    function testFuzz_CastVotesTwiceViaFlexibleVoting(
+        uint256 _proposerIndex,
+        uint256 _voterIndex,
+        uint256 _firstVote,
+        uint256 _secondVote,
+        uint256 _forVotes,
+        uint256 _againstVotes,
+        uint256 _abstainVotes
+    ) public {
+        _proposerIndex = bound(_proposerIndex, 0, _majorDelegates.length - 1);
+        _voterIndex = bound(_voterIndex, 0, _majorDelegates.length - 1);
+        uint256 _proposalId;
+
+        {
```

Comment:
I was wondering why these line were in their own block, but I can see we get a "stack too deep" if they are not.